### ISSUE TITLE
Material video playback randomly going black

### ISSUE THREAD

**Issue by H3BO3** [Frooxius/NeosPublic#588]
I'm not sure if this is one or a combination of the two recent changes made to video playback - I have a video provider in my homeworld that loops a short mp4 file from the neos cloud. It gets referenced by a few external materials stored in other locations. Every now and then, it randomly seems like the texture source gets unassigned and all the materials go dark, but inside the materials the video is still present. If I drag out the reference and drop it back in, it works again.

The other thing is that I cannot seem to make the video provider work with Unity video playback - I optimized the file for unity so I could play it with maximum compatibility/stability, but no matter how I switch the desired playback engine it still says the current one is VLC. Switching the engine tends to make the video go missing/checkerboard for a while, and it doesn't initialize or play until after a short period.

**Comment by Frooxius** [comment]
Can you check what happens with those things referencing the material that goes black? Do the references break? What if you pull the material from them, can the video texture be seen correctly?

**Comment by H3BO3** [comment]
The video is still present in the materials and playing in the slot. If I pull a reference for it from the material slot I can see the video in the reference as well. If I drop it back into the same slot it starts playing on the material again until it breaks the next time.

**Comment by Frooxius** [comment]
That sounds odd. How does the reference change when you drop it back in? Where exactly are you picking it from? 

Do you perhaps have a video of this in action?

**Comment by H3BO3** [comment]
It doesn't change. I just pull it, drop it back in and it works on the material itself again. I am pulling the reference from the emission slot of the material, which references the video provider. Dropping it back into the same slot I pulled it from seems to "refresh" it. I do not have a video but I can get one later if more information is needed.

**Comment by H3BO3** [comment]
https://streamable.com/0hmxwm First video about the referenced video textures breaking. It breaks existing video previews in inspectors too until explicitly refreshed.

https://streamable.com/2hxdx4 Second video about not being able to force Unity as the playback engine instead of VLC. I can tell it's not using Unity because I get the VLC initialization hitch sometimes when I start the world.

**Comment by Frooxius** [comment]
@H3BO3 Hmm that almost looks like the playback engine changes the texture in the background. I'll try to add a check and mitigation for that. 

For the video not being able to force Unity playback, can you give me the URL? Unity playback only supports very narrow set of formats and if the system doesn't think it'll play it, it won't work.

**Comment by Frooxius** [comment]
Ok I've added a mechanism by which the change of video texture should auto-propagate to materials when it happens on Unity's end. This should fix it hopefully. Reopen if it doesn't!

**Comment by shiftyscales** [comment]
Video textures in a VLC player will now go black consistently every time after the first playback ends. Case reported by PurpleJuice. A test video can be found here:

neosdb:///52da71d8acd94fef2841de569456620e96852eae866acb5ab1e6aa5c9c11aced.mp4

**Comment by H3BO3** [comment]
neosdb:///4d774757867a3d9e6b4b1a7440b335fba762c1d6026ebef5cb2b63116d9e86f4 my mp4 file that won't play on unity now after the updates

**Comment by Frooxius** [comment]
@shiftyscales I can't replicate that, it keeps playing correctly for me. Did you replicate that behavior on your end?

@H3BO3 I see. It seems that it's excluding it because it checks the extension. The Unity Native is only allowed for mp4 files, but the Neos URL is missing extension. I can add a MIME check for this, which could help if it's that, but what's odd is that you say it happened only after update.

Are you 100 % sure it was the update? This behavior was in Neos for a very long time now, so it should've behaved the same before, given everything else being the same. Unless the mp4 got stripped from the URL.

**Comment by shiftyscales** [comment]
Yes. Every time I spawned in the copy of the video player PurpleJuice had sent me with that asset, and each time I created a new video player, and placed the URL to that asset inside- after playback completes/the source pauses, the video player goes black.

The video texture provider will properly display the asset while a reference to it is held, and playback will occur properly for one more loop if a new player is spawned by pressing primary while gripping the reference to spawn out another video player.

But reliably, every time after the first playback, the video remains black.

Were you using the asset provided and playing it back with libVLC? This does not seem to occur when playback is done through Unity Native.

**Comment by H3BO3** [comment]
Well that's weird, pretty sure I didn't touch the neosdb extension. I'm not sure why it would have gotten stripped, but I was under the impression that checking playability would be more intelligent than that, so MIME checks would be a good idea.

**Comment by Frooxius** [comment]
I've looked into it. I added a system which fetches the MIME from the file or from Neos database. However there's an issue, the Unity Native playback doesn't accept the file if it doesn't have extension.

When you originally imported the file, it was referenced on your system directly, which included the extension (even after it gets uploaded, Neos just updates the associated URLs). However if you've moved, deleted or changed it, it needed to redownload it from the cloud, into the cache.

Problem is, in the cache it doesn't have any extension, so Unity Native will reject it even if it's perfectly capable of playing it. It's kinda stupid reason, but there's no nice way around it. People have been asking for ability to do that for years from what I can find.

However, with the next update you'll be able to stream it from the Neos cloud with Unity Native, since http(s) URL's don't care about extension.
